api/envoy/config/cluster/v3/cluster.proto

```diff
@@ -1222,6 +1222,10 @@ message Cluster {
   // If ``connection_pool_per_downstream_connection`` is true, the cluster will use a separate
   // connection pool for every downstream connection
   bool connection_pool_per_downstream_connection = 51;
+
+  // If ``ignore_removal`` is true, the cluster manager will ignore the cluster while removing during CDS event.
+  // This is useful when cluster is created and managed by custom implementation. For example, Dynamic Forward Proxy.
```

Comment:
I'm not sure this is the right place to make this change.
Shouldn't this be part of the dynamic-forward-proxy config, and not per-cluster?
The risk is that clusters are will linger for a long time, and it is unclear why would one want to keep them around with no way to remove.

Reply:
Thanks. That's a valid question. Yes, this can be part of dynamic-forward-proxy, and we can instruct it to recreate the DFP cluster if it's not found (in case it was removed by the cluster manager).

However, in environments like service-mesh or configurations where CDS can be a very frequent event, the cluster manager will end up removing it very frequently. Also, when used as an egress, DFPClusters can be high in number, and removing and recreating them on CDS after the cluster manager removes them can affect performance.

In use cases like dynamic-forward-proxy, where clusters are created and managed by a custom implementation (create/update/remove), CDS should not impact them. dynamic-forward-proxy has TTL which does remove cluster if not used for the TTL duration. 

We can hide the field from docs and let this to be used for clusters managed by custom implementations.

Open for feedback and thoughts.

Reply:
> hide the field from docs 

Personally, I think it could be a choice, and note clearly in comment: it's just for internal usage, not for users.

Another choice is: we add one more argument, `ignore_removal`, in `ClusterManagerImpl::addOrUpdateCluster` and `ClusterManagerImpl::loadCluster`.
It should works too, but with more code base changes.

Reply:
Right! @doujiang24 The later one needs more changes. Leaving note and hide from docs is simpler. 
Made changes: Note in the comment and hide it from docs. 

Open for thoughts

Reply:
Hiding from the docs should not be considered, IMHO, as it should be used for API that needs more testing (not to be hidden from some use-cases).
I think the question is whether it makes sense for DFP to have both clusters that are removed and clusters that are kept all the time. From your comments I gather that in DFP one would want to ignore the CDS removal, so it seems that a single knob (on the DFP config) makes more sense, no?

Reply:
On CDS event when clusters are being removed, the cluster manager cannot differentiate whether it is a DFP cluster or a normal cluster. 

There is nothing to differentiate and prevent these DFP clusters from getting removed by cluster manager.

Please correct me if I'm wrong or not making sense.

Reply:
The lifecycle of the sub clusters are managed by the DFP cluster, both creation and deletion.
So, we want to add a flag for the sub clusters to not be removed by other events.

The flag do not have to be a configuration, just a internal flag.

If hidding from docs is not be considered, maybe we can try the option 2？ https://github.com/envoyproxy/envoy/pull/35848#discussion_r1759645687 cc @itsLucario

Reply:
Right! Summarizing the other approach. 

We'll remove the `ignore_removal` attribute addition from Cluster API Spec.
Introduce an optional argument `ignore_removal` default set to `false` (optional param to ensure existing flow won't break) in `ClusterManagerImpl:: addOrUpdateCluster` and `ClusterManagerImpl::loadCluster` so that ClusterData can set this value and use can check while removing.

Does this sound right! If people are fine, I'll proceed with working on changes.

Reply:
FWIW, I didn't say that removing the API field is required, I just suggested moving it to the DFP config instead of having it in a per-Cluster config.

Taking a step back, there are a few things to clarify:
1. Changing the API should be done in a considerate manner, as it impacts not only Envoy users, but also xDS-server maintainers. Moreover, removing fields essentially cannot be done, so we prefer to get it mostly right now. That is why we are discussing if a field should be added, and if so where.
2. If a PR suggests changing the way Envoy behaves right now (breaking behavior change), this can either be done by adding a field or a runtime-flag. The former is a long-term support, say a filter that needs to support 2 modes of operation. The latter is when a bug is observed, and the fix causes a user-facing change. Runtime-flags allow Envoy users ~6 months to gradually ensure that the other systems behave correctly with the new change.

In the context of this PR, it is unclear to me whether it should be a config-enabled feature or a runtime-flag guarded change. I'm leaning towards the runtime guarded change according to your description (but could be persuaded towards an API feature if there are use-cases where the clusters that are added by DFP need to removed upon a CDS update).

Reply:
Thanks for the thoughts!
1. Totally agreed.  Chaging API should be done in considerate manner. After your thoughts it makes more sense to follow different approach. I feel the approach mentioned by @doujiang24 is fiesible and easy to use even in future usecases.
2. The PR is not going to change any current behaviour and doesn't break any current behaviour as well. This should be straight enough that we want to avoid removal of cluster by cluster manager whose lifecycle is managed by custom implementation.

I have updated the description accordingly.